// @langchain/langgraph-checkpoint-redis/dist/index.js
import { escapeRediSearchTagValue } from "./utils.js";
import { WRITE_KEYS_ZSET_PREFIX } from "./constants.js";
import { BaseCheckpointSaver, TASKS, copyCheckpoint, maxChannelVersion, uuid6 } from "@langchain/langgraph-checkpoint";
import { createClient, createCluster } from "redis";
//#region src/index.ts
const SCHEMAS = [
	{
		index: "checkpoints",
		prefix: "checkpoint:",
		schema: {
			"$.thread_id": {
				type: "TAG",
				AS: "thread_id"
			},
			"$.checkpoint_ns": {
				type: "TAG",
				AS: "checkpoint_ns"
			},
			"$.checkpoint_id": {
				type: "TAG",
				AS: "checkpoint_id"
			},
			"$.parent_checkpoint_id": {
				type: "TAG",
				AS: "parent_checkpoint_id"
			},
			"$.checkpoint_ts": {
				type: "NUMERIC",
				AS: "checkpoint_ts"
			},
			"$.has_writes": {
				type: "TAG",
				AS: "has_writes"
			},
			"$.source": {
				type: "TAG",
				AS: "source"
			},
			"$.step": {
				type: "NUMERIC",
				AS: "step"
			}
		}
	},
	{
		index: "checkpoint_blobs",
		prefix: "checkpoint_blob:",
		schema: {
			"$.thread_id": {
				type: "TAG",
				AS: "thread_id"
			},
			"$.checkpoint_ns": {
				type: "TAG",
				AS: "checkpoint_ns"
			},
			"$.checkpoint_id": {
				type: "TAG",
				AS: "checkpoint_id"
			},
			"$.channel": {
				type: "TAG",
				AS: "channel"
			},
			"$.version": {
				type: "TAG",
				AS: "version"
			},
			"$.type": {
				type: "TAG",
				AS: "type"
			}
		}
	},
	{
		index: "checkpoint_writes",
		prefix: "checkpoint_write:",
		schema: {
			"$.thread_id": {
				type: "TAG",
				AS: "thread_id"
			},
			"$.checkpoint_ns": {
				type: "TAG",
				AS: "checkpoint_ns"
			},
			"$.checkpoint_id": {
				type: "TAG",
				AS: "checkpoint_id"
			},
			"$.task_id": {
				type: "TAG",
				AS: "task_id"
			},
			"$.idx": {
				type: "NUMERIC",
				AS: "idx"
			},
			"$.channel": {
				type: "TAG",
				AS: "channel"
			},
			"$.type": {
				type: "TAG",
				AS: "type"
			}
		}
	}
];
var RedisSaver = class RedisSaver extends BaseCheckpointSaver {
	client;
	ttlConfig;
	constructor(client, ttlConfig) {
		super();
		this.client = client;
		this.ttlConfig = ttlConfig;
	}
	static async fromUrl(url, ttlConfig) {
		const client = createClient({ url });
		await client.connect();
		const saver = new RedisSaver(client, ttlConfig);
		await saver.ensureIndexes();
		return saver;
	}
	static async fromCluster(rootNodes, ttlConfig) {
		const client = createCluster({ rootNodes });
		await client.connect();
		const saver = new RedisSaver(client, ttlConfig);
		await saver.ensureIndexes();
		return saver;
	}
	async get(config) {
		return (await this.getTuple(config))?.checkpoint;
	}
	async getTuple(config) {
		const threadId = config.configurable?.thread_id;
		const checkpointNs = config.configurable?.checkpoint_ns ?? "";
		const checkpointId = config.configurable?.checkpoint_id;
		if (!threadId) return;
		let key;
		let jsonDoc;
		if (checkpointId) {
			key = `checkpoint:${threadId}:${checkpointNs}:${checkpointId}`;
			jsonDoc = await this.client.json.get(key);
		} else {
			const pattern = `checkpoint:${threadId}:${checkpointNs}:*`;
			const keys = await this.client.keys(pattern);
			if (keys.length === 0) return;
			keys.sort();
			key = keys[keys.length - 1];
			jsonDoc = await this.client.json.get(key);
		}
		if (!jsonDoc) return;
		if (this.ttlConfig?.refreshOnRead && this.ttlConfig?.defaultTTL) await this.applyTTL(key);
		const { checkpoint, pendingWrites } = await this.loadCheckpointWithWrites(jsonDoc);
		return await this.createCheckpointTuple(jsonDoc, checkpoint, pendingWrites);
	}
	async put(config, checkpoint, metadata, newVersions) {
		await this.ensureIndexes();
		const threadId = config.configurable?.thread_id;
		const checkpointNs = config.configurable?.checkpoint_ns ?? "";
		const parentCheckpointId = config.configurable?.checkpoint_id;
		if (!threadId) throw new Error("thread_id is required");
		const checkpointId = checkpoint.id || uuid6(0);
		const key = `checkpoint:${threadId}:${checkpointNs}:${checkpointId}`;
		const storedCheckpoint = copyCheckpoint(checkpoint);
		if (storedCheckpoint.channel_values && newVersions !== void 0) if (Object.keys(newVersions).length === 0) storedCheckpoint.channel_values = {};
		else {
			const filteredChannelValues = {};
			for (const channel of Object.keys(newVersions)) if (channel in storedCheckpoint.channel_values) filteredChannelValues[channel] = storedCheckpoint.channel_values[channel];
			storedCheckpoint.channel_values = filteredChannelValues;
		}
		const zsetKey = `${WRITE_KEYS_ZSET_PREFIX}:${threadId}:${checkpointNs}:${checkpointId}`;
		const writesExist = await this.client.exists(zsetKey);
		const jsonDoc = {
			thread_id: threadId,
			checkpoint_ns: checkpointNs === "" ? "__empty__" : checkpointNs,
			checkpoint_id: checkpointId,
			parent_checkpoint_id: parentCheckpointId || null,
			checkpoint: storedCheckpoint,
			metadata,
			checkpoint_ts: Date.now(),
			has_writes: writesExist ? "true" : "false"
		};
		this.addSearchableMetadataFields(jsonDoc, metadata);
		await this.client.json.set(key, "$", jsonDoc);
		if (this.ttlConfig?.defaultTTL) await this.applyTTL(key);
		return { configurable: {
			thread_id: threadId,
			checkpoint_ns: checkpointNs,
			checkpoint_id: checkpointId
		} };
	}
	async *list(config, options) {
		await this.ensureIndexes();
		if (options?.filter !== void 0) {
			const hasNullFilter = Object.values(options.filter).some((v) => v === null);
			const queryParts = [];
			if (config?.configurable?.thread_id) {
				const threadId = escapeRediSearchTagValue(config.configurable.thread_id);
				queryParts.push(`(@thread_id:{${threadId}})`);
			}
			if (config?.configurable?.checkpoint_ns !== void 0) {
				const checkpointNs = config.configurable.checkpoint_ns;
				if (checkpointNs === "") queryParts.push(`(@checkpoint_ns:{__empty__})`);
				else {
					const escapedNs = escapeRediSearchTagValue(checkpointNs);
					queryParts.push(`(@checkpoint_ns:{${escapedNs}})`);
				}
			}
			if (!options?.before && options?.filter) {
				for (const [key, value] of Object.entries(options.filter)) if (value === void 0) {} else if (value === null) {} else if (typeof value === "string") {
					const escapedKey = escapeRediSearchTagValue(key);
					const escapedValue = escapeRediSearchTagValue(value);
					queryParts.push(`(@${escapedKey}:{${escapedValue}})`);
				} else if (typeof value === "number") {
					const escapedKey = escapeRediSearchTagValue(key);
					queryParts.push(`(@${escapedKey}:[${value} ${value}])`);
				} else if (typeof value === "object" && Object.keys(value).length === 0) {}
			}
			if (queryParts.length === 0) queryParts.push("*");
			const query = queryParts.join(" ");
			const limit = options?.limit ?? 10;
			try {
				const fetchLimit = options?.before && !config?.configurable?.thread_id ? 1e3 : options?.before ? limit * 10 : limit;
				let documents = (await this.client.ft.search("checkpoints", query, {
					LIMIT: {
						from: 0,
						size: fetchLimit
					},
					SORTBY: {
						BY: "checkpoint_ts",
						DIRECTION: "DESC"
					}
				})).documents;
				let yieldedCount = 0;
				for (const doc of documents) {
					if (yieldedCount >= limit) break;
					if (options?.before?.configurable?.checkpoint_id) {
						if (doc.value.checkpoint_id >= options.before.configurable.checkpoint_id) continue;
					}
					const jsonDoc = doc.value;
					let matches = true;
					if ((hasNullFilter || options?.before) && options?.filter) {
						for (const [filterKey, filterValue] of Object.entries(options.filter)) if (filterValue === null) {
							if (jsonDoc.metadata?.[filterKey] !== null) {
								matches = false;
								break;
							}
						} else if (filterValue !== void 0) {
							const metadataValue = jsonDoc.metadata?.[filterKey];
							if (typeof filterValue === "object" && filterValue !== null) {
								if (deterministicStringify(metadataValue) !== deterministicStringify(filterValue)) {
									matches = false;
									break;
								}
							} else if (metadataValue !== filterValue) {
								matches = false;
								break;
							}
						}
						if (!matches) continue;
					}
					const { checkpoint, pendingWrites } = await this.loadCheckpointWithWrites(jsonDoc);
					yield await this.createCheckpointTuple(jsonDoc, checkpoint, pendingWrites);
					yieldedCount++;
				}
				return;
			} catch (error) {
				if (error.message?.includes("no such index")) {} else throw error;
			}
			if (config?.configurable?.thread_id) {
				const threadId = config.configurable.thread_id;
				const checkpointNs = config.configurable.checkpoint_ns ?? "";
				const pattern = `checkpoint:${threadId}:${checkpointNs}:*`;
				const keys = await this.client.keys(pattern);
				keys.sort().reverse();
				let filteredKeys = keys;
				if (options?.before?.configurable?.checkpoint_id) {
					const beforeKey = `checkpoint:${options.before.configurable.thread_id || threadId}:${options.before.configurable.checkpoint_ns ?? checkpointNs}:${options.before.configurable.checkpoint_id}`;
					const beforeIndex = keys.indexOf(beforeKey);
					if (beforeIndex > 0) filteredKeys = keys.slice(beforeIndex + 1);
					else if (beforeIndex === 0) filteredKeys = [];
				}
				const limit = options?.limit ?? 10;
				const limitedKeys = filteredKeys.slice(0, limit);
				for (const key of limitedKeys) {
					const jsonDoc = await this.client.json.get(key);
					if (jsonDoc) {
						let matches = true;
						for (const [filterKey, filterValue] of Object.entries(options.filter)) {
							const metadataValue = jsonDoc.metadata?.[filterKey];
							if (filterValue === null) {
								if (metadataValue !== null) {
									matches = false;
									break;
								}
							} else if (metadataValue !== filterValue) {
								matches = false;
								break;
							}
						}
						if (!matches) continue;
						const { checkpoint, pendingWrites } = await this.loadCheckpointWithWrites(jsonDoc);
						yield await this.createCheckpointTuple(jsonDoc, checkpoint, pendingWrites);
					}
				}
			} else {
				const globalPattern = config?.configurable?.checkpoint_ns !== void 0 ? `checkpoint:*:${config.configurable.checkpoint_ns === "" ? "__empty__" : config.configurable.checkpoint_ns}:*` : "checkpoint:*";
				const allKeys = await this.client.keys(globalPattern);
				const allDocuments = [];
				for (const key of allKeys) {
					const jsonDoc = await this.client.json.get(key);
					if (jsonDoc) allDocuments.push({
						key,
						doc: jsonDoc
					});
				}
				allDocuments.sort((a, b) => b.doc.checkpoint_ts - a.doc.checkpoint_ts);
				let yieldedCount = 0;
				const limit = options?.limit ?? 10;
				for (const { doc: jsonDoc } of allDocuments) {
					if (yieldedCount >= limit) break;
					if (options?.before?.configurable?.checkpoint_id) {
						if (jsonDoc.checkpoint_id >= options.before.configurable.checkpoint_id) continue;
					}
					let matches = true;
					if (options?.filter) {
						for (const [filterKey, filterValue] of Object.entries(options.filter)) if (filterValue === null) {
							if (jsonDoc.metadata?.[filterKey] !== null) {
								matches = false;
								break;
							}
						} else if (filterValue !== void 0) {
							const metadataValue = jsonDoc.metadata?.[filterKey];
							if (typeof filterValue === "object" && filterValue !== null) {
								if (deterministicStringify(metadataValue) !== deterministicStringify(filterValue)) {
									matches = false;
									break;
								}
							} else if (metadataValue !== filterValue) {
								matches = false;
								break;
							}
						}
						if (!matches) continue;
					}
					const { checkpoint, pendingWrites } = await this.loadCheckpointWithWrites(jsonDoc);
					yield await this.createCheckpointTuple(jsonDoc, checkpoint, pendingWrites);
					yieldedCount++;
				}
			}
			return;
		}
		const searchOptions = {
			...options,
			filter: {}
		};
		yield* this.list(config, searchOptions);
	}
	async putWrites(config, writes, taskId) {
		await this.ensureIndexes();
		const threadId = config.configurable?.thread_id;
		const checkpointNs = config.configurable?.checkpoint_ns ?? "";
		const checkpointId = config.configurable?.checkpoint_id;
		if (!threadId || !checkpointId) throw new Error("thread_id and checkpoint_id are required");
		const writeKeys = [];
		const baseTimestamp = performance.now() * 1e3;
		for (let idx = 0; idx < writes.length; idx++) {
			const [channel, value] = writes[idx];
			const writeKey = `checkpoint_write:${threadId}:${checkpointNs}:${checkpointId}:${taskId}:${idx}`;
			writeKeys.push(writeKey);
			const writeDoc = {
				thread_id: threadId,
				checkpoint_ns: checkpointNs,
				checkpoint_id: checkpointId,
				task_id: taskId,
				idx,
				channel,
				type: typeof value === "object" ? "json" : "string",
				value,
				timestamp: baseTimestamp,
				global_idx: baseTimestamp + idx
			};
			await this.client.json.set(writeKey, "$", writeDoc);
		}
		if (writeKeys.length > 0) {
			const zsetKey = `${WRITE_KEYS_ZSET_PREFIX}:${threadId}:${checkpointNs}:${checkpointId}`;
			const zaddArgs = {};
			writeKeys.forEach((key, idx) => {
				zaddArgs[key] = baseTimestamp + idx;
			});
			await this.client.zAdd(zsetKey, Object.entries(zaddArgs).map(([key, score]) => ({
				score,
				value: key
			})));
			if (this.ttlConfig?.defaultTTL) await this.applyTTL(...writeKeys, zsetKey);
		}
		const checkpointKey = `checkpoint:${threadId}:${checkpointNs}:${checkpointId}`;
		if (await this.client.exists(checkpointKey)) {
			const currentDoc = await this.client.json.get(checkpointKey);
			if (currentDoc) {
				currentDoc.has_writes = "true";
				await this.client.json.set(checkpointKey, "$", currentDoc);
			}
		}
	}
	async deleteThread(threadId) {
		const checkpointPattern = `checkpoint:${threadId}:*`;
		const checkpointKeys = await this.client.keys(checkpointPattern);
		if (checkpointKeys.length > 0) await this.client.del(checkpointKeys);
		const writesPattern = `writes:${threadId}:*`;
		const writesKeys = await this.client.keys(writesPattern);
		if (writesKeys.length > 0) await this.client.del(writesKeys);
	}
	async end() {
		await this.client.quit();
	}
	async loadPendingWrites(threadId, checkpointNs, checkpointId) {
		const pattern = `checkpoint_write:${threadId}:${checkpointNs}:${checkpointId}:*`;
		const writeKeys = await this.client.keys(pattern);
		if (writeKeys.length === 0) return;
		const writeDocuments = [];
		for (const writeKey of writeKeys) {
			const writeDoc = await this.client.json.get(writeKey);
			if (writeDoc) writeDocuments.push(writeDoc);
		}
		writeDocuments.sort((a, b) => (a.global_idx || 0) - (b.global_idx || 0));
		const pendingWrites = [];
		for (const writeDoc of writeDocuments) {
			const deserializedValue = await this.serde.loadsTyped("json", JSON.stringify(writeDoc.value));
			pendingWrites.push([
				writeDoc.task_id,
				writeDoc.channel,
				deserializedValue
			]);
		}
		return pendingWrites;
	}
	async loadCheckpointWithWrites(jsonDoc) {
		const checkpoint = await this.serde.loadsTyped("json", JSON.stringify(jsonDoc.checkpoint));
		if (checkpoint.v < 4 && jsonDoc.parent_checkpoint_id != null) {
			const actualNs = jsonDoc.checkpoint_ns === "__empty__" ? "" : jsonDoc.checkpoint_ns;
			await this.migratePendingSends(checkpoint, jsonDoc.thread_id, actualNs, jsonDoc.parent_checkpoint_id);
		}
		let pendingWrites;
		if (jsonDoc.has_writes === "true") {
			const actualNs = jsonDoc.checkpoint_ns === "__empty__" ? "" : jsonDoc.checkpoint_ns;
			pendingWrites = await this.loadPendingWrites(jsonDoc.thread_id, actualNs, jsonDoc.checkpoint_id);
		}
		return {
			checkpoint,
			pendingWrites
		};
	}
	async migratePendingSends(checkpoint, threadId, checkpointNs, parentCheckpointId) {
		const parentWrites = await this.loadPendingWrites(threadId, checkpointNs, parentCheckpointId);
		if (!parentWrites || parentWrites.length === 0) return;
		const taskWrites = parentWrites.filter(([, channel]) => channel === TASKS);
		if (taskWrites.length === 0) return;
		const allTasks = [];
		for (const [, , value] of taskWrites) allTasks.push(value);
		checkpoint.channel_values ??= {};
		checkpoint.channel_values[TASKS] = allTasks;
		checkpoint.channel_versions[TASKS] = Object.keys(checkpoint.channel_versions).length > 0 ? maxChannelVersion(...Object.values(checkpoint.channel_versions)) : 1;
	}
	async createCheckpointTuple(jsonDoc, checkpoint, pendingWrites) {
		const checkpointNs = jsonDoc.checkpoint_ns === "__empty__" ? "" : jsonDoc.checkpoint_ns;
		const metadata = await this.serde.loadsTyped("json", JSON.stringify(jsonDoc.metadata));
		return {
			config: { configurable: {
				thread_id: jsonDoc.thread_id,
				checkpoint_ns: checkpointNs,
				checkpoint_id: jsonDoc.checkpoint_id
			} },
			checkpoint,
			metadata,
			parentConfig: jsonDoc.parent_checkpoint_id ? { configurable: {
				thread_id: jsonDoc.thread_id,
				checkpoint_ns: checkpointNs,
				checkpoint_id: jsonDoc.parent_checkpoint_id
			} } : void 0,
			pendingWrites
		};
	}
	addSearchableMetadataFields(jsonDoc, metadata) {
		if (!metadata) return;
		if ("source" in metadata) jsonDoc.source = metadata.source;
		if ("step" in metadata) jsonDoc.step = metadata.step;
		if ("writes" in metadata) jsonDoc.writes = typeof metadata.writes === "object" ? JSON.stringify(metadata.writes) : metadata.writes;
		if ("score" in metadata) jsonDoc.score = metadata.score;
	}
	async applyTTL(...keys) {
		if (!this.ttlConfig?.defaultTTL) return;
		const ttlSeconds = Math.floor(this.ttlConfig.defaultTTL * 60);
		const results = await Promise.allSettled(keys.map((key) => this.client.expire(key, ttlSeconds)));
		for (let i = 0; i < results.length; i++) if (results[i].status === "rejected") console.warn(`Failed to set TTL for key ${keys[i]}:`, results[i].reason);
	}
	async ensureIndexes() {
		for (const schema of SCHEMAS) try {
			await this.client.ft.create(schema.index, schema.schema, {
				ON: "JSON",
				PREFIX: schema.prefix
			});
		} catch (error) {
			if (!error.message?.includes("Index already exists")) console.error(`Failed to create index ${schema.index}:`, error.message);
		}
	}
};
function deterministicStringify(obj) {
	if (obj === null || typeof obj !== "object") return JSON.stringify(obj);
	if (Array.isArray(obj)) return JSON.stringify(obj.map((item) => deterministicStringify(item)));
	const sortedObj = {};
	const sortedKeys = Object.keys(obj).sort();
	for (const key of sortedKeys) sortedObj[key] = obj[key];
	return JSON.stringify(sortedObj, (_, value) => {
		if (value !== null && typeof value === "object" && !Array.isArray(value)) {
			const sorted = {};
			const keys = Object.keys(value).sort();
			for (const k of keys) sorted[k] = value[k];
			return sorted;
		}
		return value;
	});
}
//#endregion
export { RedisSaver };
