// @langchain/langgraph-checkpoint-redis/dist/shallow.js
import { escapeRediSearchTagValue } from "./utils.js";
import { WRITE_KEYS_ZSET_PREFIX } from "./constants.js";
import { BaseCheckpointSaver, uuid6 } from "@langchain/langgraph-checkpoint";
import { createClient } from "redis";
//#region src/shallow.ts
function deterministicStringify(obj) {
	if (obj === null || typeof obj !== "object") return JSON.stringify(obj);
	if (Array.isArray(obj)) return JSON.stringify(obj.map((item) => deterministicStringify(item)));
	const sortedObj = {};
	const sortedKeys = Object.keys(obj).sort();
	for (const key of sortedKeys) sortedObj[key] = obj[key];
	return JSON.stringify(sortedObj, (_, value) => {
		if (value !== null && typeof value === "object" && !Array.isArray(value)) {
			const sorted = {};
			const keys = Object.keys(value).sort();
			for (const k of keys) sorted[k] = value[k];
			return sorted;
		}
		return value;
	});
}
const SCHEMAS = [{
	index: "checkpoints",
	prefix: "checkpoint:",
	schema: {
		"$.thread_id": {
			type: "TAG",
			AS: "thread_id"
		},
		"$.checkpoint_ns": {
			type: "TAG",
			AS: "checkpoint_ns"
		},
		"$.checkpoint_id": {
			type: "TAG",
			AS: "checkpoint_id"
		},
		"$.parent_checkpoint_id": {
			type: "TAG",
			AS: "parent_checkpoint_id"
		},
		"$.checkpoint_ts": {
			type: "NUMERIC",
			AS: "checkpoint_ts"
		},
		"$.has_writes": {
			type: "TAG",
			AS: "has_writes"
		},
		"$.source": {
			type: "TAG",
			AS: "source"
		},
		"$.step": {
			type: "NUMERIC",
			AS: "step"
		}
	}
}, {
	index: "checkpoint_writes",
	prefix: "checkpoint_write:",
	schema: {
		"$.thread_id": {
			type: "TAG",
			AS: "thread_id"
		},
		"$.checkpoint_ns": {
			type: "TAG",
			AS: "checkpoint_ns"
		},
		"$.checkpoint_id": {
			type: "TAG",
			AS: "checkpoint_id"
		},
		"$.task_id": {
			type: "TAG",
			AS: "task_id"
		},
		"$.idx": {
			type: "NUMERIC",
			AS: "idx"
		},
		"$.channel": {
			type: "TAG",
			AS: "channel"
		},
		"$.type": {
			type: "TAG",
			AS: "type"
		}
	}
}];
/**
* ShallowRedisSaver - A Redis checkpoint saver that only keeps the latest checkpoint per thread.
*
* This is a memory-optimized variant that:
* - Only stores the most recent checkpoint for each thread
* - Stores channel values inline (no separate blob storage)
* - Automatically cleans up old checkpoints and writes when new ones are added
* - Reduces storage usage for applications that don't need checkpoint history
*/
var ShallowRedisSaver = class ShallowRedisSaver extends BaseCheckpointSaver {
	client;
	ttlConfig;
	constructor(client, ttlConfig) {
		super();
		this.client = client;
		this.ttlConfig = ttlConfig;
	}
	static async fromUrl(url, ttlConfig) {
		const client = createClient({ url });
		await client.connect();
		const saver = new ShallowRedisSaver(client, ttlConfig);
		await saver.ensureIndexes();
		return saver;
	}
	async get(config) {
		return (await this.getTuple(config))?.checkpoint;
	}
	async put(config, checkpoint, metadata, _newVersions) {
		await this.ensureIndexes();
		const threadId = config.configurable?.thread_id;
		const checkpointNs = config.configurable?.checkpoint_ns ?? "";
		const parentCheckpointId = config.configurable?.checkpoint_id;
		if (!threadId) throw new Error("thread_id is required");
		const checkpointId = checkpoint.id || uuid6(0);
		const key = `checkpoint:${threadId}:${checkpointNs}:shallow`;
		let prevCheckpointData = null;
		let prevCheckpointId = null;
		try {
			prevCheckpointData = await this.client.json.get(key);
			if (prevCheckpointData && typeof prevCheckpointData === "object") prevCheckpointId = prevCheckpointData.checkpoint_id;
		} catch (error) {}
		if (prevCheckpointId && prevCheckpointId !== checkpointId) await this.cleanupOldCheckpoint(threadId, checkpointNs, prevCheckpointId);
		const checkpointCopy = {
			...checkpoint,
			channel_values: checkpoint.channel_values || {},
			channel_blobs: void 0
		};
		const zsetKey = `${WRITE_KEYS_ZSET_PREFIX}:${threadId}:${checkpointNs}:${checkpointId}`;
		const writesExist = await this.client.exists(zsetKey);
		const jsonDoc = {
			thread_id: threadId,
			checkpoint_ns: checkpointNs,
			checkpoint_id: checkpointId,
			parent_checkpoint_id: parentCheckpointId || null,
			checkpoint: checkpointCopy,
			metadata: this.sanitizeMetadata(metadata),
			checkpoint_ts: Date.now(),
			has_writes: writesExist ? "true" : "false"
		};
		this.addSearchableMetadataFields(jsonDoc, metadata);
		await this.client.json.set(key, "$", jsonDoc);
		if (this.ttlConfig?.defaultTTL) await this.applyTTL(key);
		return { configurable: {
			thread_id: threadId,
			checkpoint_ns: checkpointNs,
			checkpoint_id: checkpointId
		} };
	}
	async getTuple(config) {
		const threadId = config.configurable?.thread_id;
		const checkpointNs = config.configurable?.checkpoint_ns ?? "";
		const checkpointId = config.configurable?.checkpoint_id;
		if (!threadId) return;
		const key = `checkpoint:${threadId}:${checkpointNs}:shallow`;
		const jsonDoc = await this.client.json.get(key);
		if (!jsonDoc) return;
		if (checkpointId && jsonDoc.checkpoint_id !== checkpointId) return;
		if (this.ttlConfig?.refreshOnRead && this.ttlConfig?.defaultTTL) await this.applyTTL(key);
		const checkpoint = await this.serde.loadsTyped("json", JSON.stringify(jsonDoc.checkpoint));
		let pendingWrites;
		if (jsonDoc.has_writes === "true") pendingWrites = await this.loadPendingWrites(jsonDoc.thread_id, jsonDoc.checkpoint_ns, jsonDoc.checkpoint_id);
		return await this.createCheckpointTuple(jsonDoc, checkpoint, pendingWrites);
	}
	async *list(config, options) {
		await this.ensureIndexes();
		if (config?.configurable?.thread_id) {
			const tuple = await this.getTuple(config);
			if (tuple) if (options?.filter) {
				if (this.checkMetadataFilterMatch(tuple.metadata, options.filter)) yield tuple;
			} else yield tuple;
		} else {
			const queryParts = [];
			if (options?.filter) {
				for (const [key, value] of Object.entries(options.filter)) if (value === void 0) {} else if (value === null) {} else if (typeof value === "string") {
					const escapedKey = escapeRediSearchTagValue(key);
					const escapedValue = escapeRediSearchTagValue(value);
					queryParts.push(`(@${escapedKey}:{${escapedValue}})`);
				} else if (typeof value === "number") {
					const escapedKey = escapeRediSearchTagValue(key);
					queryParts.push(`(@${escapedKey}:[${value} ${value}])`);
				}
			}
			if (queryParts.length === 0) queryParts.push("*");
			const query = queryParts.join(" ");
			const limit = options?.limit ?? 10;
			try {
				const results = await this.client.ft.search("checkpoints", query, {
					LIMIT: {
						from: 0,
						size: limit * 2
					},
					SORTBY: {
						BY: "checkpoint_ts",
						DIRECTION: "DESC"
					}
				});
				const seenThreads = /* @__PURE__ */ new Set();
				let yieldCount = 0;
				for (const doc of results.documents) {
					if (yieldCount >= limit) break;
					const jsonDoc = doc.value;
					const threadKey = `${jsonDoc.thread_id}:${jsonDoc.checkpoint_ns}`;
					if (seenThreads.has(threadKey)) continue;
					seenThreads.add(threadKey);
					if (options?.filter) {
						if (!this.checkMetadataFilterMatch(jsonDoc.metadata, options.filter)) continue;
					}
					const checkpoint = await this.serde.loadsTyped("json", JSON.stringify(jsonDoc.checkpoint));
					yield await this.createCheckpointTuple(jsonDoc, checkpoint);
					yieldCount++;
				}
			} catch (error) {
				if (error.message?.includes("no such index")) {
					const keys = await this.client.keys(`checkpoint:*:*:shallow`);
					if (keys.length === 0) return;
					keys.sort().reverse();
					const seenThreads = /* @__PURE__ */ new Set();
					let yieldCount = 0;
					const limit = options?.limit ?? 10;
					for (const key of keys) {
						if (yieldCount >= limit) break;
						const jsonDoc = await this.client.json.get(key);
						if (!jsonDoc) continue;
						const threadKey = `${jsonDoc.thread_id}:${jsonDoc.checkpoint_ns}`;
						if (seenThreads.has(threadKey)) continue;
						seenThreads.add(threadKey);
						if (options?.filter) {
							if (!this.checkMetadataFilterMatch(jsonDoc.metadata, options.filter)) continue;
						}
						const checkpoint = await this.serde.loadsTyped("json", JSON.stringify(jsonDoc.checkpoint));
						yield await this.createCheckpointTuple(jsonDoc, checkpoint);
						yieldCount++;
					}
					return;
				}
				throw error;
			}
		}
	}
	async putWrites(config, writes, taskId) {
		await this.ensureIndexes();
		const threadId = config.configurable?.thread_id;
		const checkpointNs = config.configurable?.checkpoint_ns ?? "";
		const checkpointId = config.configurable?.checkpoint_id;
		if (!threadId || !checkpointId) throw new Error("thread_id and checkpoint_id are required");
		const writePattern = `checkpoint_write:${threadId}:${checkpointNs}:${checkpointId}:${taskId}:*`;
		const oldWriteKeys = await this.client.keys(writePattern);
		if (oldWriteKeys.length > 0) await this.client.del(oldWriteKeys);
		const writeKeys = [];
		for (let idx = 0; idx < writes.length; idx++) {
			const [channel, value] = writes[idx];
			const writeKey = `checkpoint_write:${threadId}:${checkpointNs}:${checkpointId}:${taskId}:${idx}`;
			writeKeys.push(writeKey);
			const writeDoc = {
				thread_id: threadId,
				checkpoint_ns: checkpointNs,
				checkpoint_id: checkpointId,
				task_id: taskId,
				idx,
				channel,
				type: typeof value === "object" ? "json" : "string",
				value
			};
			await this.client.json.set(writeKey, "$", writeDoc);
		}
		if (writeKeys.length > 0) {
			const zsetKey = `${WRITE_KEYS_ZSET_PREFIX}:${threadId}:${checkpointNs}:${checkpointId}`;
			const zaddArgs = {};
			writeKeys.forEach((key, idx) => {
				zaddArgs[key] = idx;
			});
			await this.client.zAdd(zsetKey, Object.entries(zaddArgs).map(([key, score]) => ({
				score,
				value: key
			})));
			if (this.ttlConfig?.defaultTTL) await this.applyTTL(...writeKeys, zsetKey);
		}
		const checkpointKey = `checkpoint:${threadId}:${checkpointNs}:shallow`;
		if (await this.client.exists(checkpointKey)) {
			const currentDoc = await this.client.json.get(checkpointKey);
			if (currentDoc) {
				currentDoc.has_writes = "true";
				await this.client.json.set(checkpointKey, "$", currentDoc);
			}
		}
	}
	async deleteThread(threadId) {
		const checkpointPattern = `checkpoint:${threadId}:*:shallow`;
		const checkpointKeys = await this.client.keys(checkpointPattern);
		if (checkpointKeys.length > 0) await this.client.del(checkpointKeys);
		const writesPattern = `checkpoint_write:${threadId}:*`;
		const writesKeys = await this.client.keys(writesPattern);
		if (writesKeys.length > 0) await this.client.del(writesKeys);
		const zsetPattern = `${WRITE_KEYS_ZSET_PREFIX}:${threadId}:*`;
		const zsetKeys = await this.client.keys(zsetPattern);
		if (zsetKeys.length > 0) await this.client.del(zsetKeys);
	}
	async end() {
		await this.client.quit();
	}
	addSearchableMetadataFields(jsonDoc, metadata) {
		if (!metadata) return;
		if ("source" in metadata) jsonDoc.source = metadata.source;
		if ("step" in metadata) jsonDoc.step = metadata.step;
		if ("writes" in metadata) jsonDoc.writes = typeof metadata.writes === "object" ? JSON.stringify(metadata.writes) : metadata.writes;
		if ("score" in metadata) jsonDoc.score = metadata.score;
	}
	async createCheckpointTuple(jsonDoc, checkpoint, pendingWrites) {
		const metadata = await this.serde.loadsTyped("json", JSON.stringify(jsonDoc.metadata));
		return {
			config: { configurable: {
				thread_id: jsonDoc.thread_id,
				checkpoint_ns: jsonDoc.checkpoint_ns,
				checkpoint_id: jsonDoc.checkpoint_id
			} },
			checkpoint,
			metadata,
			parentConfig: jsonDoc.parent_checkpoint_id ? { configurable: {
				thread_id: jsonDoc.thread_id,
				checkpoint_ns: jsonDoc.checkpoint_ns,
				checkpoint_id: jsonDoc.parent_checkpoint_id
			} } : void 0,
			pendingWrites
		};
	}
	async applyTTL(...keys) {
		if (!this.ttlConfig?.defaultTTL) return;
		const ttlSeconds = Math.floor(this.ttlConfig.defaultTTL * 60);
		const results = await Promise.allSettled(keys.map((key) => this.client.expire(key, ttlSeconds)));
		for (let i = 0; i < results.length; i++) if (results[i].status === "rejected") console.warn(`Failed to set TTL for key ${keys[i]}:`, results[i].reason);
	}
	async loadPendingWrites(threadId, checkpointNs, checkpointId) {
		const zsetKey = `${WRITE_KEYS_ZSET_PREFIX}:${threadId}:${checkpointNs}:${checkpointId}`;
		const writeKeys = await this.client.zRange(zsetKey, 0, -1);
		if (writeKeys.length === 0) return;
		const pendingWrites = [];
		for (const writeKey of writeKeys) {
			const writeDoc = await this.client.json.get(writeKey);
			if (writeDoc) {
				const deserializedValue = await this.serde.loadsTyped("json", JSON.stringify(writeDoc.value));
				pendingWrites.push([
					writeDoc.task_id,
					writeDoc.channel,
					deserializedValue
				]);
			}
		}
		return pendingWrites;
	}
	checkMetadataFilterMatch(metadata, filter) {
		for (const [key, value] of Object.entries(filter)) {
			const metadataValue = metadata?.[key];
			if (value === null) {
				if (!(key in (metadata || {})) || metadataValue !== null) return false;
			} else if (typeof value === "object" && !Array.isArray(value)) {
				if (typeof metadataValue !== "object" || metadataValue === null) return false;
				if (deterministicStringify(value) !== deterministicStringify(metadataValue)) return false;
			} else if (metadataValue !== value) return false;
		}
		return true;
	}
	async cleanupOldCheckpoint(threadId, checkpointNs, oldCheckpointId) {
		const writePattern = `checkpoint_write:${threadId}:${checkpointNs}:${oldCheckpointId}:*`;
		const oldWriteKeys = await this.client.keys(writePattern);
		if (oldWriteKeys.length > 0) await this.client.del(oldWriteKeys);
		const zsetKey = `${WRITE_KEYS_ZSET_PREFIX}:${threadId}:${checkpointNs}:${oldCheckpointId}`;
		await this.client.del(zsetKey);
		const blobPattern = `checkpoint_blob:${threadId}:${checkpointNs}:${oldCheckpointId}:*`;
		const oldBlobKeys = await this.client.keys(blobPattern);
		if (oldBlobKeys.length > 0) await this.client.del(oldBlobKeys);
	}
	sanitizeMetadata(metadata) {
		if (!metadata) return {};
		const sanitized = {};
		for (const [key, value] of Object.entries(metadata)) {
			const sanitizedKey = key.replace(/\x00/g, "");
			sanitized[sanitizedKey] = typeof value === "string" ? value.replace(/\x00/g, "") : value;
		}
		return sanitized;
	}
	async ensureIndexes() {
		for (const schema of SCHEMAS) try {
			await this.client.ft.create(schema.index, schema.schema, {
				ON: "JSON",
				PREFIX: schema.prefix
			});
		} catch (error) {
			if (!error.message?.includes("Index already exists")) console.error(`Failed to create index ${schema.index}:`, error.message);
		}
	}
};
//#endregion
export { ShallowRedisSaver };
